### PR TITLE
[dims] extend wb_dims() to accept multiple columns

### LINKED ISSUE #843: wb_dims: Error in wb_dims(rows = 2:(nrow(data_in) + 1), cols = sel_columns) :    You must supply positive, consecutive values to `cols`
Hello! 
I would like to apply conditional formatting to selected columns (across all rows), but those columns are not next to each other. E.g. the index is:
```
> sel_columns
[1]  6  9 12 15 18 21 24 27
```
When I try: 
```
wb$add_conditional_formatting(my_var, dims = wb_dims(rows = 2:(nrow(data_in)+1), cols = sel_columns), 
rule = ">=2", style = "negative")
```
the error is: 
```
Error in wb_dims(rows = 2:(nrow(data_in) + 1), cols = sel_columns) : 
 You must supply positive, consecutive values to `cols`
```

As my simple workaround, I can loop across all elements and apply conditional_formating as many times as needed:
```
for (i in sel_columns){
   wb$add_conditional_formatting(my_var, dims = wb_dims(rows = 2:(nrow(data_in)+1), cols = i), 
   rule = ">=2", style = "negative")
```
But I am wondering if am I  just missing something and that option is already implemented?
Thanks! 

openxlsx2 version 1.1

### PR DESCRIPTION
**Pull request by JanMarvin**
Fix #843

This PR extends `wb_dims()` to accept multiple columns:

``` r
library(openxlsx2)
wb <- wb_workbook() %>% 
  wb_add_worksheet() %>% 
  wb_add_data(x = mtcars) %>% 
  wb_add_fill(dims = wb_dims(x = mtcars, cols = c("cyl", "gear"), select = "col_names"), color = wb_color("blue")) %>% 
  wb_add_fill(dims = wb_dims(x = mtcars, cols = c("cyl", "gear")), color = wb_color("orange"))

if (interactive())  wb %>% wb_open()
```

Screenshot 2024-05-18 at 13 54 04

``` r
# just to show what is going on
wb_dims(x = mtcars, cols = c("cyl"))
#> [1] "B2:B33"

wb_dims(x = mtcars, cols = c("cyl", "gear"))
#> [1] "B2:B33,J2:J33"

wb_dims(x = mtcars, select = "col_names")
#> [1] "A1:K1"

wb_dims(x = mtcars, cols = c("cyl", "gear"), select = "col_names")
#> [1] "B1,J1"
```

### CODE REVIEW

**Comment by JanMarvin** [comment]
Related issue: https://github.com/JanMarvin/openxlsx2/issues/796#issuecomment-1737463448

**Comment by olivroy** [comment]
Thanks for caring care of this! I'd want to take my time on this review to get it as right as possible on this PR. Hope this is not too time-sensitive?

**Comment by JanMarvin** [comment]
Sure, take your time. I've adjusted the `wb_dims()` code, but let's put it this way, I'm not entirely satisfied myself. The `wb_dims()` code is a bit strange to me. I was just happy, when I finally had it working. Ideally, I would have wanted the `rowcol_to_dim()`/`rowcol_to_dims()` functions to take whatever we put into it and return the requested dimensions -cols is a start, but rows would be nice too-  I just doubt I have the energy to rewrite the existing code.

For the second part, I had some issues with `row_names = TRUE` and stumbled over it. It's nothing critical, I was just confused.

**Comment by olivroy** [comment]
I may not have time to review before end of week..

**Comment by JanMarvin** [comment]
No worries, it’s not time critical.

**Comment by JanMarvin** [comment]
Achievement unlocked! Make it even more confusing 😄 

``` r
library(openxlsx2)
d_full      <- wb_dims(x = mtcars)
d_col       <- wb_dims(x = mtcars, cols = c(2))
d_cols      <- wb_dims(x = mtcars, cols = c(2, 4))
d_row       <- wb_dims(x = mtcars, rows = c(4))
d_rows      <- wb_dims(x = mtcars, rows = c(2, 4:5))
d_cols_rows <- wb_dims(x = mtcars, cols = c(2, 4), rows = c(2, 4:5))

wb <- wb_workbook()$
  add_worksheet()$add_data(x = mtcars)$add_fill(dims = d_full, color = wb_color("green"))$
  add_worksheet()$add_data(x = mtcars)$add_fill(dims = d_col,  color = wb_color("green"))$
  add_worksheet()$add_data(x = mtcars)$add_fill(dims = d_cols, color = wb_color("green"))$
  add_worksheet()$add_data(x = mtcars)$add_fill(dims = d_row,  color = wb_color("green"))$
  add_worksheet()$add_data(x = mtcars)$add_fill(dims = d_rows, color = wb_color("green"))$
  add_worksheet()$add_data(x = mtcars)$add_fill(dims = d_cols_rows, color = wb_color("green"))

if (interactive()) wb$open()
```

File: R/utils.R [hunk 1]
```diff
@@ -904,15 +902,20 @@ wb_dims <- function(..., select = NULL) {
...
     }
 
-    if (!is.null(rows_arg)) {
+    if (!is.null(rows_arg)) { # && length(rows_arg) == 1L && all(diff(rows_arg) == 1L)) {
       if (min(rows_arg) > 1) {
-        frow <- frow + min(rows_arg) - 1L
+        if (all(diff(rows_arg) == 1L))
+          frow <- frow + min(rows_arg) - 1L
+        else
+          frow         <- vapply(rows_arg, function(x) frow + min(x) - 1L, NA_real_)
```
Comment (olivroy) [inline R/utils.R:918]:
```suggestion
          frow <- vapply(rows_arg, function(x) frow + min(x) - 1L, NA_real_)
```

**Review by olivroy — APPROVED** [review]
Sorry to have kept you waiting! I think this is a nice improvement. You were able to simplify the code from the previous time I looked at this?

I guess that a good way to test this would be to see if this works with various functions.

`,` will work with `wb_add_conditional_formatting()`, data? 


I think it is worth publicizing this elsewhere in docs (in a follow-up perhaps?)

Thankfully, I didn't have to work with Excel a lot lately ;).

Maybe we can ask user testing too at this point? I remember this issue in particular that would be solved with this 

https://github.com/JanMarvin/openxlsx2/issues/843

**Comment by olivroy** [comment]
After merge, I will install this version so I can get a bit of integration testing before hand!

**Comment by JanMarvin** [comment]
> Sorry to have kept you waiting! I think this is a nice improvement. You were able to simplify the code from the previous time I looked at this?

We have all the time in the world! But the code is not really simplified. I only fixed what I broke with non consecutive rows. And I'm not really hyped on the coding in this PR 😅 

> I guess that a good way to test this would be to see if this works with various functions.
> 
> `,` will work with `wb_add_conditional_formatting()`, data?

Haven't checked this. Chances are it wont work. Afaik conditional formatting requires a consecutive range, but haven't tried.

> I think it is worth publicizing this elsewhere in docs (in a follow-up perhaps?)
> 
> Thankfully, I didn't have to work with Excel a lot lately ;).

Yeah, that's a luxury I don't have in my life ... at least with `openxlsx2` I don't have to many chores in that nasty spreadsheet software.

> Maybe we can ask user testing too at this point? I remember this issue in particular that would be solved with this
> 
> #843

With open source development everyone wants working releases and nobody wants to test. Same reason R-Core has to do many patch releases ...

**Comment by JanMarvin** [comment]
Alright, merging this one. After all we can always revert and at least our tests don't bark at us

**Pull request merged**